Persist state machine

HammadB opened the pull request:
## Description of changes

*Summarize the changes made by this PR.* 
 - Improvements & Bug fixes
	 - Adds a state machine for persistence tests.

## Test plan
These are the tests.

## Documentation Changes
None required. The existing documentation covers persist-on-exit behavior.

HammadB commented:
Ready for re-review post api changes

HammadB commented on chromadb/api/local.py:
```diff
@@ -24,6 +24,7 @@
 from chromadb.telemetry import Telemetry
```
The primary change in this file is to add persist() to all the write path methods. I would love verification that none are missed.

HammadB replied:
Update: no longer relevant since we don't persist()

HammadB commented on chromadb/db/duckdb.py:
```diff
@@ -440,14 +463,15 @@ def load(self):
 
     def __del__(self):
         logger.info("PersistentDuckDB del, about to run persist")
-        self.persist()
```
This is the only non-linter change to duckdb.py. We remove __del__ calls to persist. 

Our use of __del__ to call persist is fundamentally wrong.

1. “It is not guaranteed that __del__() methods are called for objects that still exist when the interpreter exits.” This means that you will not reliably persist on exit. In my test it also means that the subprocess that loads and checks can do odd things. 
2. In many contexts, there are hidden references to the object that are unreliable - leading to buggy code that assumes persist on exist or that del on that object cleans it up. 
3. __del__ masks errors. As according to the documentation - dependent resources in the object may have been cleaned up in an unpredictable order. I’ve seen several transient instances of duckdb errors inside __del__ that I can’t explain but this explains it. “__del__() can be executed during interpreter shutdown. As a consequence, the global variables it needs to access (including other modules) may already have been deleted or set to None. “

HammadB replied:
https://docs.python.org/3/reference/datamodel.html?highlight=__del__#object.__del__

HammadB replied:
I have found an alternative to persist on write and del, which is to use https://docs.python.org/3/library/atexit.html

HammadB commented on chromadb/test/property/test_persist.py:
```diff
@@ -1,21 +1,37 @@
     configuration = request.param
-    return lambda: chromadb.Client(configuration)
+    yield lambda: chromadb.Client(configuration)
+    save_path = configuration.persist_directory
```
Add cleanup for persistence

HammadB commented on chromadb/db/clickhouse.py:
```diff
@@ -100,7 +108,8 @@ def _delete_index(self, collection_id):
     #  UTILITY METHODS
     #
     def persist(self):
-        raise NotImplementedError("Clickhouse is a persistent database, this method is not needed")
```
This should not throw an error but should no-op.

atroyn commented on chromadb/test/property/test_persist.py:
```diff
@@ -1,21 +1,37 @@
+
+
+@pytest.fixture(scope="module", params=persist_configurations())
+def settings(request) -> Settings:
```
Use settings in `test_persist`

HammadB commented on chromadb/db/duckdb.py:
```diff
@@ -440,14 +463,15 @@ def load(self):
             )
 
     def __del__(self):
         logger.info("PersistentDuckDB del, about to run persist")
```
We cannot - since the base class has a __del__ that deletes the indices. I added a comment explaining this

levand reviewed (approved):
The code for this looks ok and I don't have any additional questions. I agree that correctness > performance for right now.

However, I do have some concern for how this is going to affect the developer experience, since it will affect performance quite substantially depending on usage patterns.. I'm imagining a case where a user is adding thousands/millions of embeddings by calling `add` repeatedly and passing embeddings one at a time.

This was never recommended or the right way to do things. But it's possible that this change could convert some existing apps from "Chroma works" to "Chroma is intolerably slow". 

Maybe we should do a quick benchmark of that pathological case just so we know what happens, before we release this?

nit: IMO, we should have separate PRs for large amounts of formatting changes, to make it easier to focus on substance.

HammadB commented:
> The code for this looks ok and I don't have any additional questions. I agree that correctness > performance for right now.
> 
> However, I do have some concern for how this is going to affect the developer experience, since it will affect performance quite substantially depending on usage patterns.. I'm imagining a case where a user is adding thousands/millions of embeddings by calling `add` repeatedly and passing embeddings one at a time.
> 
> This was never recommended or the right way to do things. But it's possible that this change could convert some existing apps from "Chroma works" to "Chroma is intolerably slow".
> 
> Maybe we should do a quick benchmark of that pathological case just so we know what happens, before we release this?
> 
> nit: IMO, we should have separate PRs for large amounts of formatting changes, to make it easier to focus on substance.

I have the same concern but there isn't a better way to do this I can think of. It is certainly slower, if we knew the exact amount what subsequent steps would you take? I think the correct action is to land the refactor without explicit persist calls with a disk-based db. 

Agree - with the nit. We need to run the formatter over the whole repo and have everyone committing code with the formatter as per #288

HammadB commented:
@levand @atroyn 

I've changed this PR to use atexit in favor of __del__ to get persist-on-exit behavior (when safely exited) and instead of persisting every write (due to perf concerns). I think this is the correct stopgap.

Additionally, I changed the persist state machine to run minimum every 10 iterations instead of at the will of hypothesis.

HammadB commented:
> 👍 on `atexit`, looks like the right stopgap solution for us.
> 
> Just verifying the intent behind `last_persist_delay`... the goal here is simply prevent it from hitting the `persist` state repeatedly without any actual changes in the data set?
> 
> I just wanted to check that we're not using state transitions as a proxy for clock time between persists since that would end up unreliable.

Yeah exactly - the persist calls are expensive so we don't want to allow hypothesis to just repeatedly call it. The goal is to force hypothesis to make changes to the data before re-persisting.
